gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java

```diff
@@ -514,6 +545,62 @@ private static ExecutorSupplier getPCUExecutorSupplier(ExecutorService pCUExecut
         : ExecutorSupplier.useExecutor(pCUExecutorService);
   }
 
+  @Override
+  public VectoredIOResult readVectored(
+      List<VectoredIORange> ranges, IntFunction<ByteBuffer> allocate, BlobId blobId)
+      throws IOException, ExecutionException, InterruptedException, TimeoutException {
+    logger.atFiner().log("readVectored() called for BlobId=%s", blobId.toString());
+    long clientInitializationDurationStartTime = System.currentTimeMillis();
+    AtomicInteger totalBytesRead = new AtomicInteger();
+    try (BlobReadSession blobReadSession =
+        storage.blobReadSession(blobId).get(30, TimeUnit.SECONDS)) {
+      long clientInitializationDuration =
+          System.currentTimeMillis() - clientInitializationDurationStartTime;
+      logger.atFiner().log("Client Initialization successful in %d", clientInitializationDuration);
+      Map<VectoredIORange, ApiFuture<byte[]>> futures =
+          ranges.stream()
+              .collect(
+                  Collectors.toMap(
+                      Function.identity(),
+                      range ->
+                          blobReadSession.readAs(
```

Comment:
From https://github.com/googleapis/java-storage/blob/main/google-cloud-storage/src/main/java/com/google/cloud/storage/ReadProjectionConfigs.java#L99

It seems like sdk will use a memory space to hold on to the data and then we are copying it back the memory space provided by the client via allocate. Am I correct?

What about us the using `ReadAsChannel` and pass of the array from the allocate?

Reply:
@singhravidutt Is correct, the bytes will accumulate in netty's pool until the future resolves, at which point they will be released after creating a single contiguous `byte[]` on the heap. Until the future resolves, no bytes will be held on the gc managed heap of the process.

Using a channel would reduce the accumulation in netty's pool (one in progress message + one pending message) would be the maximum buffer depth, but you would need a thread to read the bytes through the channel.

If you want to avoid the `byte[]` allocation, you can use the `asFutureByteString()` config. I'll add another comment with a diff showing how I'd write this.

Reply:
I think there are a couple of things being discussed here:

1) Channel vs ByteString:
I think both of these are efficient in terms of space as they will just use the netty's pool as compared to byte[] which is using additional heap space as well.

2) ByteString vs Byte[]:
My ultimate aim was to use ByteString only.
I initially used byte[] to benchmark its memory usage against ByteString, as a prior discussion with Ben suggested there would be a negligible performance difference. Now that I'm confident we have addressed potential lifecycle management concerns with ByteString, particularly around unintentionally retaining Netty memory, I have updated the implementation to use it.